File: Day33/Part2/k_alternating_subsequence_xor.cpp
#include <bits/stdc++.h>
using namespace std;

static const int BITS = 31; // for values up to < 2^31

struct Basis {
    array<int,BITS> b{};
    void insert(int x){
        for(int i=BITS-1;i>=0;--i){
            if(!((x>>i)&1)) continue;
            if(!b[i]){ b[i]=x; return; }
            x ^= b[i];
        }
    }
    void merge(const Basis& other){
        for(int v: other.b) if(v) insert(v);
    }
    int max_xor() const {
        int res=0;
        for(int i=BITS-1;i>=0;--i){
            if(b[i] && (res^b[i])>res) res ^= b[i];
        }
        return res;
    }
};

int k_alternating_max_xor(const vector<int>& nums, int k){
    int n = (int)nums.size();
    vector<Basis> B(n + k + 1); // padding
    for(int i=n-1;i>=0;--i){
        Basis cur = B[i+1]; // skip i
        Basis choose = B[i+k];
        choose.insert(nums[i]);
        cur.merge(choose);
        B[i] = cur;
    }
    return B[0].max_xor();
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout << k_alternating_max_xor(vector<int>{3,10,5,25,2,8},2) << "\n"; // Expect 26
    return 0;
}

File: Day33/Part2/k_alternating_subsequence_xor.py
from typing import List

BITS = 31  # for up to 1e9 < 2^30

def insert_basis(basis: List[int], x: int) -> None:
    for i in reversed(range(BITS)):
        if not (x >> i) & 1:
            continue
        if basis[i] == 0:
            basis[i] = x
            return
        x ^= basis[i]

def merge_basis(dst: List[int], src: List[int]) -> None:
    for v in src:
        if v:
            insert_basis(dst, v)

def max_xor(basis: List[int]) -> int:
    res = 0
    for i in reversed(range(BITS)):
        if basis[i] and (res ^ basis[i]) > res:
            res ^= basis[i]
    return res

def k_alternating_max_xor(nums: List[int], k: int) -> int:
    n = len(nums)
    # Store bases B[i]; each basis is length BITS
    B = [ [0]*BITS for _ in range(n+ k + 1) ]  # extra padding to avoid bounds
    for i in range(n-1, -1, -1):
        # Start from B[i+1] (skip i case)
        cur = B[i][:]
        merge_basis(cur, B[i+1])
        # Include choosing i then jump to i+k
        choose_basis = B[i+k][:]
        insert_basis(choose_basis, nums[i])
        merge_basis(cur, choose_basis)
        B[i] = cur
    return max_xor(B[0])

if __name__ == "__main__":
    print(k_alternating_max_xor([3,10,5,25,2,8], 2))  # Expect 26 per problem example
